---
title: 'Session 3 | Part 1 : Gene expression analysis, control/treatment experiment'
author: "Tomás Moyano"
date: "5/25/2020"
output: rmarkdown::github_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/Documentos/pcamejo/")
```

### **Watch this [video](https://drive.google.com/file/d/1EdYCPAKTPI-Z28fyUF8ri1KzDw3nazgd/view) for a more detailed explanation of the following pipeline**

During this session, we will discover genes differentially expressed when Arabidopsis' shoot is grown with no nitrate (KCl treatment), compared to its development with 5mM of nitrate. The samples analyzed here were taken after 120 min of growth.

Open R Studio (or just R) and load the following libraries.

```{r message=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DESeq2")
BiocManager::install("edgeR")
BiocManager::install("ViSEAGO")
BiocManager::install("Rgraphviz")
BiocManager::install("topGO")

library(DESeq2)
library(edgeR)
library(knitr)
library(dplyr)
library(ViSEAGO)


```

> Don't forget to set the folder where you downloaded your files as the working directory.

```{r warning=FALSE}
setwd( "~/Documentos/pcamejo/")
getwd()
```

###  Importing and formatting data.

Start importing counts table and metadata associated to the samples (previously downloaded from [Data](https://github.com/ibioChile/Transcriptomics-R-Workshop-public/tree/master/Session2-Treatment_and_Multivariate/Data) folder).

```{r}
counts <- read.table("fc0.original.counts.session2-1.txt")
metadata <- read.table("metadata_session2-1.txt", header=TRUE)
kable(head(metadata))
```

We will fix the header of counts leaving only the sample ID.

```{r}
colnames(counts) <- sapply(strsplit(colnames(counts),".",fixed=TRUE), `[`, 1)
kable(head(counts))
```

Let's order samples in metadata by Tissue, Treatment and Time. Then order samples in counts table according to metadata.

```{r}
metadata_s <- metadata %>% arrange(Tissue, Treatment,Time)
counts <- counts[,metadata_s$Sample]
dim(counts)
```

>This table has 6 samples and 32,833 genes.

Now, we will evaluate two different methods to determine differentially expressed genes (DEGs) among treatment and control experiments: [DESeq2](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8) and [EdgeR](https://academic.oup.com/bioinformatics/article/26/1/139/182458).


As input, the DESeq2 and EdgeR package expects count data as obtained, e.g., from RNA-seq or another high-throughput sequencing experiment, in the form of a matrix of integer values. The count data are presented as a table which reports, for each sample, the number of sequence fragments that have been assigned to each gene. An important analysis question is the quantification and statistical inference of systematic changes between conditions, as compared to within-condition variability.

### 1. DEseq2


The package DESeq2 provides methods to test for differential expression by use of negative binomial generalized linear models; the estimates of dispersion and logarithmic fold changes incorporate data-driven prior distributions

In this case, we will compare the treatment and control data.

First, we will create an object describing the treatment condition ('KNO3' if grown with nitrate or 'KCl' if grown with KCl instead) associated with each column in the "counts" table.

```{r}
coldata <- data.frame(row.names = metadata$Sample, condition = metadata$Treatment)
coldata
```

The function *DESeqDataSetFromMatrix* can be used if you have a matrix of read counts prepared (as we have from *RSubread*). With the count matrix, ```counts```, and the sample information, ```coldata```, we can construct a DESeqDataSet:

```{r}
dds <- DESeqDataSetFromMatrix( countData=counts,  colData=coldata, design=~condition)
dds
```

While it is not necessary to pre-filter low count genes before running the DESeq2 functions, there are two reasons which make pre-filtering useful: by removing rows in which there are very few reads, we reduce the memory size of the ```dds``` data object, and we increase the speed of the transformation and testing functions within DESeq2. Here we perform a minimal pre-filtering to keep only rows that have at least 10 reads total. Note that more strict filtering to increase power is automatically applied via independent filtering on the mean of normalized counts within the results function. This initial filter can help you use less memory and get faster results.

```{r}
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
```

By default, R will choose a reference level for factors based on alphabetical order. Then, if you never tell the DESeq2 functions which level you want to compare against (e.g. which level represents the control group), the comparisons will be based on the alphabetical order of the levels ("KCl" first, then "KNO3"). There are two solutions: you can either explicitly tell results which comparison to make using the contrast argument (this will be shown later), or you can explicitly set the factors levels. Setting the factor levels can be done in two ways, either using factor:

```{r}
dds$condition <- factor(dds$condition, levels = c("KCl","KNO3"))
```

…or using relevel, just specifying the reference level:

```{r}
dds$condition <- relevel(dds$condition, ref = "KCl")
```

The standard differential expression analysis steps are wrapped into a single function, ```DESeq```. These steps include normalization by the size of the libraries, and an adjustment in the dispersion of the data specific to the DESeq2 package (see http://bioconductor.org/packages/release/bioc/html/DESeq2.html). For more information about how this function works, we recommend you to check the manual page for ```?DESeq``` and the Methods section of the DESeq2 publication (Love, Huber, and Anders 2014). Results tables are generated using the function ```results```, which generates a table with log2 fold changes, p-values and adjusted p-values.


```{r}
dds <- DESeq(dds)
res01 <- results(dds, alpha=0.01)
summary(res01)
```

We can plot the normalized count for each gene. In this exampe we will plot the smallest adjusted p-value results.

```{r}
plotCounts(dds, gene=which.min(res01$padj), intgroup="condition")

```

Now, we can filter the results by the desired thresholds of fold change level and adjusted p-values. In this case, we will select genes with log2-fold changes (FC) greater than 1 (genes that doubled its expression) and adjusted p-values < 0.01. We will store filtered data under the ```filter``` object:

```{r}
res <- na.exclude(as.data.frame(res01))
filter <- res[(abs(res$log2FoldChange)>1 & res$padj<0.01),]
kable(head(filter))

```

```{r}
nrow(filter)
```

According to this analysis, there are 2,536 DEGs among treatment conditions at time 120 min. A table with DEGs information can be exported as a text file for further analysis.

```{r}
write.table(filter,"regulated_DESEQ2_log2FC1_padj0.01.txt", quote=F,sep="\t",  col.names = NA)
```


### 2. EdgeR

Another widely used tool for differential expression analysis is edgeR. This program uses a different method than DeSeq2, and therefore, results may be different. The different results are not mutually exclusive, since no statistical modeling can fully capture biological phenomena. Statistical methods all rely on assumptions and requirements that are only partially satisfied. Therefore, we always recommend to try a couple of methods when analyzing this type of data.

In the previous step we load the counts and metadata file into a DeSeq2 object. Similarly, to run edgeR, we need to combine data counts and metadata into a DGElist object:

```{r}
dge <- DGEList(counts=counts,group=metadata$Treatment)
dge
```

Similarly to the DeSeq2 pipeline, we will include a step to filter out lowly expressed gene. We will use the ```filterByExpr``` function. By default, the function keeps genes with about 10 read counts or more.


```{r}
keep <- filterByExpr(dge,group = metadata$Treatment)
dge <- dge[keep,,keep.lib.sizes=FALSE]
dge
```

>Note that the total number of genes changed compared to the previous step.

We will use the function ```calcNormFactors``` to estimate a set of scaling factors, which minimizes differences in the distribution of log-fold changes between samples. The default method for computing these scale factors uses a trimmed mean of M-values (TMM) between each pair of samples.

```{r}
dge <- calcNormFactors(dge, method = "TMM")
dge
```

To compare Control vs Treatment samples, we need a design matrix with "Treatment" conditions and to estimate the dispersion of the samples (review [Session 1](https://github.com/ibioChile/Transcriptomics-R-Workshop-public/blob/master/Session1-Temporal_Analysis/3_Pipeline-Gene_expression_analysis.md) for more information):

```{r}
design <- model.matrix(~metadata$Treatment)
dge <- estimateDisp(dge,design)
dge
```

The last step of this pipeline is to compute genewise exact tests for differences between the two groups means. EdgeR asumes a negative-binomially distribution. The function ```topTags``` extracts the most differentially expressed genes from a test object, ranked either by p-value or by absolute log-fold-change. In this case, we use this function to sort the object ```et``` by adjusted p-values.

```{r}
et <- exactTest(dge)
et.res<-topTags(et,n = dim(et$table)[1], adjust.method="BH")
kable(head(data.frame(et.res)))
```
We can see the results as a plot:


```{r}
plotMD(et,p.value = 0.01,adjust.method = "fdr")
```


Now, we will filter the results by the fold change level and adjusted p-values. Again, we will select genes wit log2FC > 1 and adjusted p-value < 0.01.

```{r}
et.res.0.01_1<-et.res$table[et.res$table$FDR<0.01 & abs(et.res$table$logFC)>1,]
kable(head(et.res.0.01_1))
```


```{r}
nrow(et.res.0.01_1)
```

According to this analysis, 2,661 genes were DE by treatment at time 120 min.

The resulting table can be written as text file for further analysis.

```{r}
write.table(et.res.0.01_1,"regulated_edgeR_log2FC1_padj0.01.txt", quote=F,sep="\t",  col.names = NA)
```


### 3. Gene comparison

Now, let's compare the results of both methods. For that, we will thera are a lot of tool to intersect lists of genes and plot. In this exam'le, we will use an "UpSet plot" with the list of DEGs reulting from both DeSeq2 and EdgeR.

```{r}

#install.packages("UpSetR")
library(UpSetR)

list.DE<-list(deseq=rownames(filter),edgeR=rownames(et.res.0.01_1))

upset(fromList(list.DE), order.by = "freq")
```

There is a great overlap among the DEGs found by the two methods. There is a large overlap between the DEG found by the two methods.  There are a few genes that are found exclusively by each method, this is because different normalization and comparison methods are used. For more information see https://pubmed.ncbi.nlm.nih.gov/22988256/.

### 4. GO enrichment.

Once a list of DEGs has been obtained, it is useful to know the biological functions in which those genes may be involved in. This can be done by using specific databases, where the function of the organisms' genes has been previosuly identified. One of the most widely used databases to describe genes is [Gene Ontology](www.geneontology.org). This database is a collaborative initiative, in which the information related to genes is stored using a structured language.

There are a lot of tools to analyze lists of genes with gene ontology data. In this example, we will utilize the ViSEAGO package of Bioconductor (https://www.bioconductor.org/packages/release/bioc/html/ViSEAGO.html). Although ViSEAGO is prepared to be quickly used for model organisms (as the case of Arabidopsis), in this example, the gene-function relation (GOterm) will be loaded from the website http://current.geneontology.org/products/pages/downloads.html (see this session's [Requirements](https://github.com/ibioChile/Transcriptomics-R-Workshop-public/blob/master/Session3-Treatment_and_Multivariate/1_Requirements.md))

Read this file:

```{r}
tair.gaf<-read.delim("tair.gaf.gz",comment.char = "!",fill=T,header=F)
kable(head(tair.gaf))
```

Let's select the columns needed to load Gene Ontology information into the database:

```{r}
library(GO.db)
goframeData<-(tair.gaf[,c(10,3,5,7)])# This is specific for this file

kable(head(goframeData))


```

```{r}
goframeData<-goframeData[as.matrix(nchar(as.matrix(goframeData[,1])))==9,] # This is specific for this file
goframeData<-cbind("Arabidopsis",goframeData)
colnames(goframeData)<-c("taxid","gene_id","gene_symbol","GOID","evidence")
goframeData<-goframeData[goframeData$GOID%in%keys(GO.db),]
write.table(goframeData,"goframeData.arabidopsis",row.names = F,sep="\t",quote=F)
kable(head(goframeData))

```


To build a customizable database for ViSEAGO, you need to run the folwwing code. The GO file is then loaded into ```myGENE2GO``` as a structured object, and the hierarchical gene ontology tree is then built from the annotations.

```{r message=FALSE}
library(ViSEAGO)

Custom<-ViSEAGO::Custom2GO(file = "goframeDAta.arabidopsis")
myGENE2GO<-ViSEAGO::annotate(  "Arabidopsis",  Custom )  

myGENE2GO

```


Then, we load the parameters to perform this analysis. For this case, we will use a DESeq2 object previously generated and select genes from the intersection of edgeR and DeSeq2. We will define as "background universe" the total list of genes being expressed in the plant. For Gene Ontology analysis, we can select data from different ontology categories. In this case, the "biological processes (BP)" category will be used.


```{r}

background <- names(dds)

selection <- intersect(rownames(filter),rownames(et.res.0.01_1))

BP<-ViSEAGO::create_topGOdata(geneSel=selection, allGenes=background,  gene2GO=myGENE2GO,   ont="BP",  nodeSize=5)

classic<-topGO::runTest(BP,  algorithm ="classic",  statistic = "fisher")

classic

```

Finally, the results can be displayed as a table or a gene ontology tree:

```{r}
# merge results
BP_sResults<-ViSEAGO::merge_enrich_terms(  Input=list(  condition=c("BP","classic")  ))

# display the merged table
```

The resulting table can be written as text file for further analyses.

```{r}
# print the merged table in a file
ViSEAGO::show_table(  BP_sResults,  "GO_intersect.deseq_edgeR_genes.BP.txt")
BP_sResults.table<-read.table("GO_intersect.deseq_edgeR_genes.BP.txt",header = T,row.names=1)

kable(head(BP_sResults.table))
```


```{r}

showSigOfNodes(GOdata = BP,score(classic) , useInfo = 'all',firstSigNodes = nrow(BP_sResults.table))

printGraph(BP, classic, firstSigNodes = nrow(BP_sResults.table), fn.prefix = "BP", useInfo = "all", pdfSW = TRUE)


```



```


