---
title: 'Session 3 | Part 2 : Gene expression analysis, multivariate experiments'
author: "Pamela Camejo"
date: "5/15/2020"
output: rmarkdown::github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/Users/pamelacamejo/Documents/IBIO/R workshop/Session2/Pipeline 2/")
```
### **Watch [this video](https://drive.google.com/file/d/1MT_04vC4c23LWJKTjEdTVf7cYj1YPdZf/view?usp=sharing) for a more detailed explanation of the following pipeline**

During this session, we will discover genes differentially expressed in the genome of Arabidopsis when considering time, treatment and tissue, as well as the interaction of these factors.

Open R Studio (or just R) and load the following libraries.

```{r message=FALSE}
library(edgeR)
library(knitr)
library(dplyr)
library(reshape2)
library(ggplot2)
require(gridExtra)
library(ComplexHeatmap)
library(circlize)
library(VennDiagram)

theme_set(theme_bw())
```

> Don't forget to set the folder where you downloaded your files as the working directory.

```{r warning=FALSE}
setwd("~/Documents/IBIO/R workshop/Session2/Pipeline 2/")
```

### 1. Importing and formatting data.

Start importing counts table and metadata associated to the samples (previously downloaded from [Data](https://github.com/ibioChile/Transcriptomics-R-Workshop-public/tree/master/Session3-Treatment_and_Multivariate/Data) folder).

```{r}
counts <- read.table("fc0.original.counts.session2-2.txt")
metadata <- read.table("metadata_session2-2.txt", header=TRUE)
kable(head(metadata))
```

We will fix the header of counts leaving only the sample ID.

```{r}
colnames(counts) <- sapply(strsplit(colnames(counts),".",fixed=TRUE), `[`, 1)
kable(head(counts))
```

Let's order samples in metadata by Tissue, Treatment and Time. Then order samples in counts table according to metadata.
```{r}
metadata_s <- metadata %>% arrange(Tissue, desc(Treatment),Time)
counts <- counts[,metadata_s$Sample]
dim(counts)
```

>This table has 114 samples and 32,833 genes.

### 2. Filter and normalization

We start the analysis by filtering out genes for which no count has been found.

```{r}
counts_filt1 <- counts[rowSums(counts) > 0, ]
dim(counts_filt1)
```

>This leave us with 31,520 genes. 

Now, we will create an object of type CountDataSet to carry out normalization:

```{r}
dge <- DGEList(counts_filt1)
```

It is often useful to visualize the count distribution, to compute “pseudo counts”, which are log-transformed counts. We will store raw pseudo-counts in a data frame for later use.

```{r}
pseudo_counts <- cpm(dge, log=TRUE, prior.count = 1)

df_raw <- melt(t(pseudo_counts), id = rownames(pseudo_counts))
names(df_raw)[1:2]<- c("sample", "id")
df_raw$method <- rep("Raw counts", nrow(df_raw))  
kable(head(df_raw))
```

The object ```df_raw``` will be used later to compare the effect of normalization over count distribution in different samples.

We will try TMM normalization and compared to raw data.

```{r}
dge1 <- calcNormFactors(dge, method = "TMM")
```
>Try different normalization methods to pick the right one.

Normalized data is then obtained by the function cpm. 

```{r}
pseudo_TMM <- cpm(dge1, log=TRUE, prior.count = 1)

df_TMM <- melt(t(pseudo_TMM), id = rownames(pseudo_TMM))
names(df_TMM)[1:2]<- c("sample", "id")
df_TMM$method <- rep("TMM", nrow(df_TMM))  
kable(head(df_TMM))
```

To compare data counts pre and post normalization, we will generate **boxplots** of each sample with counts distribution.

```{r fig1, fig.align="center", dpi=300, fig.height=4, fig.width=8}
df_allnorm <- rbind(df_raw, df_TMM)

p <- ggplot(data=df_allnorm, aes(x=sample, y=value, fill=method)) 
p <- p + geom_boxplot()  
p <- p + theme(title = element_text(size=10), axis.text.x = element_blank(), 
               axis.ticks.x = element_blank())
p <- p + ggtitle("Boxplots of normalized pseudo counts compared to raw data")
p <- p + facet_grid(method ~ .) 
p <- p + ylab(expression(log[2] ~ ("normalized cpm  0=1"))) + xlab("")

print(p)
```

We can see in this plot that one of the samples failed (very low number of reads). Let's check the total number of reads obtained by each sample.

```{r}
kable(dge1$samples)
```

It looks like sample 'SRR5440803' has an abnormal number of reads (3,000 vs. ~10^7 in the other samples). This sample correspond to **Shoot, KCl treatment at time 90, replicate 2**. Let's remove this sample from the analysis and re-run the previous code.

```{r}
counts_filt2 <- counts_filt1 %>% select(-"SRR5440803")
metadata_s2 <- metadata_s[metadata_s$Sample != "SRR5440803",]

pseudo_counts2 <- log2(counts_filt2 + 1)

df_raw2 <- melt(t(pseudo_counts2), id = rownames(pseudo_counts2))
names(df_raw2)[1:2]<- c("sample", "id")
df_raw2$method <- rep("Raw counts", nrow(df_raw2))  
kable(head(df_raw2))
```

EdgeR normalization:

```{r}
dge2 <- DGEList(counts_filt2)
dge2 <- calcNormFactors(dge2, method = "TMM")
```

We will also include a step to filter out lowly expressed gene. We will use the ```filterByExpr``` function from edgeR. This function implements the filtering strategy described by Chen et al (2016). By default, the function keeps genes with about 10 read counts or more.

```{r}
keep <- filterByExpr(dge2)
dge2 <- dge2[keep, , keep.lib.sizes=FALSE]
```

Normalized/filtered data:

```{r}
pseudo_TMM <- cpm(dge2, log=TRUE, prior.count = 1)

df_TMM <- melt(t(pseudo_TMM), id = rownames(pseudo_TMM))
names(df_TMM)[1:2]<- c("sample", "id")
df_TMM$method <- rep("TMM", nrow(df_TMM))  
kable(head(df_TMM))
```

Now, let's generate boxplots one more time.

```{r fig2, fig.align="center", dpi=300, fig.height=4, fig.width=8}
df_allnorm <- rbind(df_raw2, df_TMM)

p <- ggplot(data=df_allnorm, aes(x=sample, y=value, fill=method)) 
p <- p + geom_boxplot() 
p <- p + theme(title = element_text(size=10), axis.text.x = element_blank(), 
               axis.ticks.x = element_blank())
p <- p + ggtitle("Boxplots of normalized pseudo counts compared to raw data")
p <- p + facet_grid(method ~ .) 
p <- p + ylab(expression(log[2] ~ ("normalized cpm  0=1"))) + xlab("")

print(p)
```

Another way to visualize data distribution is by generating data density plots.

```{r fig3, fig.align="center", dpi=200, fig.height=5, fig.width=6}
p <- ggplot(data=df_allnorm, aes(x=value, colour=sample)) + geom_density()
p <- p + theme(title = element_text(size=10), legend.position = "none")
p <- p + ggtitle("Density of normalized pseudo counts compared to raw data")
p <- p + facet_grid(method ~ . ) 
p <- p + scale_color_manual(values=colorRampPalette(c("yellow2","goldenrod"))(ncol(pseudo_TMM)))
p <- p + xlab(expression(log[2] ~ ("normalized cpm  0=1"))) 

print(p)
```

By looking at these plots, we observe that now medians of each dataset are similar, that we got rid of samples with very low counts, and that data distributios are comparable between samples.

### 3. Principal Component Analisis (PCA)

Principal component analysis (PCA) is another way to examine inter-sample relationships by producing a plot. There are several functions to do a PCA analysis in R. Here we will use prcomp that comes with the default R package *stats*.

First, we will perform a PCA on the pseudo counts obtained after TMM normalization, using *prcomp*.

```{r}
expr_data_bot_pca <- prcomp(t(pseudo_TMM),center = TRUE,scale. = TRUE)
dat.pca <- data.frame(expr_data_bot_pca$x) 
percentVar <- data.frame(summary(expr_data_bot_pca)$importance)
```

We will use this information to plot a PCA using ggplot.

```{r fig4, fig.align="center", dpi=200, fig.height=4, fig.width=5}

metadata_s2$Time = as.factor(metadata_s2$Time)

ggplot(dat.pca, aes(PC1,PC2)) + geom_point(aes(shape=metadata_s2$Tissue, colour=metadata_s2$Time),size=4) + 
  scale_shape_manual(values=c(15,16)) +
  scale_colour_brewer(palette = "Set3") + 
  theme(
    axis.line = element_line(colour = "black", size = 0.5, linetype = "solid"),
    panel.background = element_rect(fill = "white", colour = "black", size = 1, linetype = "solid"),
    panel.grid.major = element_line(colour ="grey", size = , linetype = "dashed"),
    panel.grid.minor = element_line(colour ="black", size = , linetype = "dashed"),
    legend.text = element_text(size = 6),
    legend.title = element_text(color = "blue", size = 8)
  ) +
  xlab(paste0("PC1: (",round(percentVar[2,1] *100,2),"%)")) +
  ylab(paste0("PC2: (",round(percentVar[2,2] * 100,2),"%)")) + 
  coord_cartesian() +
  labs(shape = "Tissue", colour = "Time")
```

This figure shows a clear separation between samples from root and shoot, as well as samples from times 10 and 15min in the case of root.

### 4. Linear models

Linear models are among the most used statistical methods. T-tests, ANOVA, ANCOVA and regression can all be formulated as special cases of linear models. This model includes some assumptions, as we usually assume that the noise is Normally distributed. The linear model is readily extended to p predictor variables X1⋯Xp with observations xi1⋯xip. In this  case the predictors are the variables analyzed (Time, Treatment and Tissue) and the observations the normalized data quantified in each sample. 

The following script adjusts each gene data to a linear model, considering "Tissue, Treatment and Time" as variables of the model. Afterwards, it also corrects p-values using the Bonferroni method (Rice, 1989), in which the p-values are multiplied by the number of comparisons. The table *Result_lm_adj* contains adjusted p-values for each variable and gene.

```{r}
TMM_counts <- cpm(dge2)
TMM_counts_t <- t(TMM_counts) 

Result_lm <- NULL
for (i in 1:ncol(TMM_counts_t)) {
  test_lm <- lm(TMM_counts_t[,i] ~ metadata_s2$Tissue  +
                   metadata_s2$Treatment + 
                   metadata_s2$Time)
  Result_lm <- rbind(Result_lm,coef(summary(test_lm))[,4])
  }

Result_lm <- as.data.frame(Result_lm)
rownames(Result_lm) <- colnames(TMM_counts_t)

Result_lm_adj <- as.data.frame(apply(Result_lm,2, function(x) p.adjust(x, method = "bonferroni")))
kable(head(Result_lm_adj))
```

>Bonferroni corecction is a very stringent method. You could try this analysis adjusting p-values with other methods. Some of the most used are "BH" (or "fdr") and "BY". Check the usage of the *p.adjust* function for more information.

**DEGs over time**

Now, let's take a look at genes differentially expressed (DE) over time. From the linear model tested above, we obtained p-values for each to the time points evaluated when compared to the control. We will consider genes with adjusted p-values < 0.01 at any time point to be 'time-dependant DE genes'.

```{r}
DE_time_df <- pseudo_TMM[rowSums(Result_lm_adj[,4:12] < 0.01) >= 1,] #These are the columns of p-values comparing time points.
nrow(DE_time_df)
```

According to this analysis, 883 genes are DE over time.

>Additionally, you could also include a fold-change threshold to select DE genes. For instance, we could select genes whose expression changes at least 1.5 times at any time point, with respect to time 0.

We will generate a Heatmap of expression to visualize how the expression of these genes is changing over time. There are many libraries designed to generate heatmaps (including ggplot). Here, we'll use the function *Heatmap* from the package *ComplexHeatmap*. This is a very flexible and easy to use library, and it is easy to change parameters and add information. This function requires scaling the data externally.

```{r}
DE_time_df_scaled <- t(scale(t(DE_time_df))) # z-score data scaling to be able to compare among genes
```

```{r}
column_ha = HeatmapAnnotation(Tissue = metadata_s2$Tissue,
                              Treatment = metadata_s2$Treatment,
                              Time = as.numeric(as.character(metadata_s2$Time)),
                              col = list(Tissue = c("Root" = "brown", "Shoot" = "darkgreen"),
               Treatment = c("None" = "grey", "KCl" = "orange"),
               Time = colorRamp2(c(0, 120), c("white", "black")))) # we will add a top column with variables labeling
```

```{r fig5, fig.align="center", dpi=150, fig.height=10, fig.width=8}
Heatmap(DE_time_df_scaled,col=colorRampPalette(c("red","white","blue"))(100),
        name = "z-score expression",
        border ="black",cluster_rows =TRUE,cluster_columns =FALSE, show_row_names = FALSE,
        show_column_names = FALSE, top_annotation = column_ha)
```

**DEGs with treatment** 

Now, let's take a look at genes differentially expressed (DE) when Arabidopsis is grown without nitrate.

```{r}
DE_treatment_df <- pseudo_TMM[Result_lm_adj[,3] < 0.01,] # This is the column of p-values comparing treatments.
nrow(DE_treatment_df)
```

According to this analysis, 1,948 genes are DE when no nitrate is added to the media. Let's generate a heatmap with these genes now.

```{r fig6, fig.align="center", dpi=150, fig.height=10, fig.width=6}
DE_treatment_df_scaled <- t(scale(t(DE_treatment_df)))

Heatmap(DE_treatment_df_scaled,col=colorRampPalette(c("red","white","blue"))(100),
        name = "z-score expression",
        border ="black",cluster_rows =TRUE,cluster_columns =FALSE, show_row_names = FALSE,
        show_column_names = FALSE, top_annotation = column_ha)
```

We could also use boxplots of each group of samples to visualize changes of gene expression for one or more genes of interest. In the next script, we pick gene "AT1G01140" (CIPK9, CBL-INTERACTING PROTEIN KINASE 9) to  create a boxplot.

```{r}
id_1 = "AT1G01140"

DE_treatment_df1 <- DE_treatment_df[id_1,]

DE_treatment_meta_df <- cbind(exp = DE_treatment_df1, metadata_s2[,c(2:3,5)]) # We bind to the normalized table, metadata information of Tissue, Treatment and Time. 

DE_treatment_meta_melted <- melt(DE_treatment_meta_df)
```

```{r fig7, fig.align="center", dpi=150, fig.height=4, fig.width=6}
p <- DE_treatment_meta_melted %>% ggplot(aes(x=Time, y = value, colour=Treatment)) + geom_boxplot()  
p <- p + facet_grid( Tissue ~ .,scale="free_y")
p <- p + ggtitle("Boxplots of normalized pseudo counts of AT1G01140")
p <- p + ylab(expression(log[2] ~ (normalized ~ "cpm, 0=1"))) + xlab("Time")
p <- p + scale_colour_manual(values=c("red","blue"))
p <- p + theme(title = element_text(size=10))
print(p)
```
We observe in this plot that the expression of this gene in both tissues is higher when Arabidopsis is grown without nitrate.

> We could also use a line chart with the average expression per point to visualize this information.

**DEGs among tissues** 

```{r}
DE_tissue_df <- pseudo_TMM[Result_lm_adj[,2] < 0.01,] # This is the column of p-values comparing tissues.
nrow(DE_tissue_df)
```
According to this analysis, 13,706 genes have a significantly different expression in root and shoot.

Heatmap:

```{r fig8, fig.align="center", dpi=150, fig.height=7, fig.width=8}
DE_tissue_df <- t(scale(t(DE_tissue_df)))

# Heatmap
Heatmap(DE_tissue_df,col=colorRampPalette(c("red","white","blue"))(100),
        name = "z-score expression",
        border ="black",cluster_rows =TRUE,cluster_columns =FALSE, show_row_names = FALSE,
        show_column_names = FALSE, top_annotation = column_ha)
```


**Interactions**

Furthermore, we can test genes responding to the interaction of variables. For example, the following code identify genes with differential expression over time **and** when the plant grows without nitrate addition (Time and Treatment interaction). 

```{r}

Result_lm2 <- NULL
for (i in 1:ncol(TMM_counts_t)) {
  test_lm2 <- lm(TMM_counts_t[,i] ~ metadata_s2$Treatment:as.factor(metadata_s2$Time))
  Result_lm2 <- rbind(Result_lm2,coef(summary(test_lm2))[,4])
  }

Result_lm2 <- as.data.frame(Result_lm2)
rownames(Result_lm2) <- colnames(TMM_counts_t)

Result_lm_adj2 <- as.data.frame(apply(Result_lm2,2, function(x) p.adjust(x, method = "bonferroni")))
kable(head(Result_lm_adj2))
```

Now, let's take a look at genes differentially expressed (DE) over time. From the linear model tested above, we obtained p-values for each to the time points evaluated when compared to the control. We will consider genes with adjusted p-values < 0.01 at any time point to be 'time-dependant DE genes'.

```{r}
DE_int_df <- pseudo_TMM[rowSums(Result_lm_adj2[,2:19] < 0.01) >= 1,] #These are the columns of p-values comparing variables.
nrow(DE_int_df)
```

There are 1,652 genes responding to both time and treatment in Arabidopsis.

Heatmap:

```{r fig9, fig.align="center", dpi=150, fig.height=7, fig.width=8}

DE_int_df <- t(scale(t(DE_int_df)))

# Heatmap
Heatmap(DE_int_df,col=colorRampPalette(c("red","white","blue"))(100),
        name = "z-score expression",
        border ="black",cluster_rows =TRUE,cluster_columns =FALSE, show_row_names = FALSE,
        show_column_names = FALSE, top_annotation = column_ha)
```

Let's generate a boxplot of one of this DE genes, "AT1G01180", to visualize changes of gene expression:

```{r}
id_1 = "AT1G01180"

DE_treatment_df1 <- DE_treatment_df[id_1,]

DE_treatment_meta_df <- cbind(exp = DE_treatment_df1, metadata_s2[,c(2:3,5)]) # We bind to the normalized table, metadata information of Tissue, Treatment and Time. 

DE_treatment_meta_melted <- melt(DE_treatment_meta_df)
```

```{r fig10, fig.align="center", dpi=150, fig.height=4, fig.width=6}
p <- DE_treatment_meta_melted %>% ggplot(aes(x=Time, y = value, colour=Treatment)) + geom_boxplot()  
p <- p + facet_grid( Tissue ~ .)
p <- p + ggtitle("Boxplots of normalized pseudo counts of AT1G01180")
p <- p + ylab(expression(log[2] ~ (normalized ~ count + 1))) + xlab("Time")
p <- p + scale_colour_manual(values=c("red","blue"))
p <- p + theme(title = element_text(size=10))
print(p)
```
We observe in this plot that "AT1G01180" is changing its expression over time (in all experiments in Root and the plant grown under normal conditions in Shoot), but also we see the expression of this gene is greater in samples grown without nitrate (KCl). 


### 5. Generalized Linear models

Generalized linear models (GLMs) are an extension of classical linear models to nonnormally distributed response data. In this section, we fit a generalized linear model (glm) using the the functions belonging to the edgeR package (similar analysis in Section 1). 

First, we create a design matrix:

```{r}
dsn <- model.matrix(~Tissue+Treatment+Time, data=metadata_s2)
kable(head(dsn))
```

As we did in Session 2, now, we will estimate the dispersion in three steps: a common dispersion first, then a trended dispersion and finally a gene specific dispersion:

```{r}
dge2 <- estimateGLMCommonDisp(dge2, dsn)
dge2 <- estimateGLMTrendedDisp(dge2, dsn)
dge2 <- estimateGLMTagwiseDisp(dge2, dsn)
```

Finally, the model (negative binomial generalized log-linear model) is fitted with the function glmFit.

```{r}
fit <- glmFit(dge2, dsn)
```

Let's see the coefficients of this model (colnames of table):

```{r}
colnames(fit)
```

These coefficients indicate the contrasts that we can perform with this model: TreatmentNone (None vs. KCl), TissueShoot (Shoot vs. Root), and all Time coefficients are contrasted against Time 0.

**DEGs over time**

Now, let's take a look at genes differentially expressed (DE) over time. F

DEGs can be inferred by performing a likelihood ratio test (LRT) from the model with results from coefficients 4 to 12 (TimeX). As done with results from the linear model, we will adjust p-values using the Bonferroni method.

```{r}
lrt_time <- glmLRT(fit, coef=4:12)
lrt_time_adj <- cbind(lrt_time$table, p_adj = p.adjust(lrt_time$table$PValue, method = "bonferroni"))
```

Again, we will consider genes with adjusted p-values < 0.01 at any time point to be 'time-dependant DE genes'.

```{r}
lrt_time_sig <- pseudo_TMM[lrt_time_adj$p_adj < 0.01,]
nrow(lrt_time_sig)
```

According to this analysis, 9,330 genes are DE over time.

>Again, you could also include a fold-change threshold to select DE genes.

Let's compare the list of DE genes over time obtained with this method to the time-dependant genes resulted from the linear model:

```{r fig11, fig.align="center", dpi=100, fig.height=3, fig.width=3.5}
grid.newpage()
venn.plot <- draw.pairwise.venn(area1      = nrow(DE_time_df),
                                area2      = nrow(lrt_time_sig),
                                cross.area = length(intersect(rownames(DE_time_df),rownames(lrt_time_sig))),
                                category   = c("lm" , "glm"),
                                fill = c("yellow","cyan3"),
                                alpha= 0.1,
                                scaled     = TRUE)
```

It looks like the majority of DEGs from the linear model (806 out of 883) are also inferred by the glm method. However, the number of genes obtained from the glm is ~10 times greater than the list from lm. Let's take a look at the expression of these genes:

```{r}
lrt_time_sig_scaled <- t(scale(t(lrt_time_sig))) # z-score data scaling to be able to compare among genes
```

```{r fig12, fig.align="center", dpi=150, fig.height=10, fig.width=8}
Heatmap(lrt_time_sig_scaled,col=colorRampPalette(c("red","white","blue"))(100),
        name = "z-score expression",
        border ="black",cluster_rows =TRUE,cluster_columns =FALSE, show_row_names = FALSE,
        show_column_names = FALSE, top_annotation = column_ha)
```

Finally, let's see how the list of DEGs looks for Treatment and Tissue and compare it to what we found with the lm:

**DEGs with treatment**

```{r}
lrt_treat <- glmLRT(fit, coef=3)
lrt_treat_adj <- cbind(lrt_treat$table, p_adj = p.adjust(lrt_treat$table$PValue, method = "bonferroni"))
lrt_treat_sig <- pseudo_TMM[lrt_treat_adj$p_adj < 0.01,]
nrow(lrt_treat_sig)
```

According to this analysis, 2,355 genes are DE when no nitrate is added to the media. This number is more similar to the one obtained from the lm (1,948 DEGs). 

```{r fig13, fig.align="center", dpi=100, fig.height=3, fig.width=3.5}
grid.newpage()
venn.plot <- draw.pairwise.venn(area1      = nrow(DE_treatment_df),
                                area2      = nrow(lrt_treat_sig),
                                cross.area = length(intersect(rownames(DE_treatment_df),rownames(lrt_treat_sig))),
                                category   = c("lm" , "glm"),
                                fill = c("yellow","cyan3"),
                                alpha= 0.1,
                                scaled     = TRUE)
```

A large portion of these genes is shared by the two methods.


**DEGs among tissues**

```{r}
lrt_tissue <- glmLRT(fit, coef=2)
lrt_tissue_adj <- cbind(lrt_tissue$table, p_adj = p.adjust(lrt_tissue$table$PValue, method = "bonferroni"))
lrt_tissue_sig <- pseudo_TMM[lrt_tissue_adj$p_adj < 0.01,]
nrow(lrt_tissue_sig)
```

According to this analysis, 13,922 genes have a differential expression among tissues, in comparison with the 13,706 genes obtained from the lm. 

```{r fig14, fig.align="center", dpi=100, fig.height=3, fig.width=3.5}
grid.newpage()
venn.plot <- draw.pairwise.venn(area1      = nrow(DE_tissue_df),
                                area2      = nrow(lrt_tissue_sig),
                                cross.area = length(intersect(rownames(DE_tissue_df),rownames(lrt_tissue_sig))),
                                category   = c("lm" , "glm"),
                                fill = c("yellow","cyan3"),
                                alpha= 0.1,
                                scaled     = TRUE)
```

Again, a large portion of these genes is shared by the two methods.

**Interactions**

Using this method, we can also test genes responding to the interaction of Time and Treatment. 

```{r}
dsn2 <- model.matrix(~Time:Treatment, data=metadata_s2)
dsn2 <- dsn2[,c(-2,-12)] # We remove the combination Time0:TreatmentKCl, since there is not sample with this condition. Since we need a full-rank matrix, we will also remove Time0:TreatmentNone from the matrix.
kable(head(dsn2))
```

As we did in Session 2, now, we will estimate the dispersion in three steps: a common dispersion first, then a trended dispersion and finally a gene specific dispersion:

```{r}
dge2 <- estimateGLMCommonDisp(dge2, dsn2)
dge2 <- estimateGLMTrendedDisp(dge2, dsn2)
dge2 <- estimateGLMTagwiseDisp(dge2, dsn2)
fit <- glmFit(dge2, dsn2)
```

Let's see the coefficients of this model (colnames of table):

```{r}
colnames(fit)
```

All of them are interaction coefficients.

```{r}
lrt_int <- glmLRT(fit, coef=2:19)
lrt_int_adj <- cbind(lrt_int$table, p_adj = p.adjust(lrt_int$table$PValue, method = "bonferroni"))
lrt_int_sig <- pseudo_TMM[lrt_int_adj$p_adj < 0.01,]
nrow(lrt_int_sig)
```

We found 3,067 time and treatment dependant DEGs using this method (1,652 genes from lm).

```{r fig15, fig.align="center", dpi=100, fig.height=3, fig.width=3.5}
grid.newpage()
venn.plot <- draw.pairwise.venn(area1      = nrow(DE_int_df),
                                area2      = nrow(lrt_int_sig),
                                cross.area = length(intersect(rownames(DE_int_df),rownames(lrt_int_sig))),
                                category   = c("lm" , "glm"),
                                fill = c("yellow","cyan3"),
                                alpha= 0.1,
                                scaled     = TRUE)
```

A large portion of DEGs obtained from lm is included in this method.

### END ###

>In order to accurately repeat this analyis, make sure that your libraries' versions match the ones used here:

```{r}
sessionInfo()
```
