---
title: "Session3-Transcriptional_networks"
author: "tm"
date: "5/28/2020"
output: rmarkdown::github_document
---


```{r setup, include=FALSE, message=FALSE}
library(knitr)
library(dplyr)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/Documentos/pcamejo/")

```

The rapid increase in the availability of transcriptomics data represents an opportunity for biologists to integrating and interpreting these data sets. The opportunity is to use this information to generate testable hypothesis to understand molecular mechanisms controlling gene expression and biological processes. Asuccessful strategy to generate tractable hypotheses from transcriptomics data has been to buildnetwork graphs based on patterns of gene co-expression. This guide includes basic instructions for the operation of widely used open source platforms R and Cytoscape. Even though the data we used in this example was obtained from Arabidopsis thaliana, the workflow developed in this guide can be easily adapted to work with RNA-seq data from any organism. The instructions are based in this book chapter: https://pubmed.ncbi.nlm.nih.gov/29525965/



###  Importing and formatting data.

First set your working directory:

```{r}
setwd("~/Documentos/pcamejo/")
```


Start importing counts table and metadata associated to the samples (previously downloaded from [Data](https://github.com/ibioChile/Transcriptomics-R-Workshop/tree/master/Session2-Treatment_and_Multivariate/Data) folder).

```{r}
counts <- read.table("fc0.original.counts.session2-2.txt",sep="\t",header=T,row.names = 1)

kable(head(counts))
```

Normalization of gene reads counts from RNA-seq data.

Reads counts for each gene can be used to calculate correlation of gene expression . In this example, read counts should be normalized to the number of reads that were effectively mapped. In this example, we use median normalization of the data. This method can be applied using the EBSeq R package. EBSeq generates a normalized matrix using different methods including median normalization. The normalization will be applied to all the mapped reads selected in previous sections.



The following commands can perform the mentioned normalization. First install EBSeq:

```{r message=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("EBSeq")
library(EBSeq)

```


Then normalize the data in "counts" file executing EBSeq and store the normalized data in "NormData" object:

```{r}
NormData <- GetNormalizedMat(counts, MedianNorm(counts))

```


The resulting gene expression matrix "NormData" contains unique row identifiers and row counts obtained from different experiments from the previous steo in each column. After normalization, it is recommended to delete very low counts data or sum a unit to all data in order to avoid values equal to zero. Then is useful to generate a logarithmic matrix of the data to standardize the variance. All previous recommendations can be done by simply:

```{r}
NormData.log <- log2(NormData+1)
```


The "NormData.log" object stores normalized counts in logarithmic scale from all libraries.


##Calculating correlation of gene expression for each pair of genes

We are now getting close to the goal of this guide which is to generate a gene co-expression network. In this step, we will determine correlation of every pair of DEGs across the complete data set. First, the user should select DEGs obtained by DESeq2 from the "NormData.log" table generated previously and extract their normalized counts, using the following command:

```{r}

DEGs.deseq2<-read.table("regulated_DESEQ2_log2FC1_padj0.01.txt",header=T,row.names=1)

select<-DEGs.deseq2[abs(DEGs.deseq2$log2FoldChange)>1,]

# For this example, we will use a stricter criterion and we will select the genes that change at least 4 times with respect to the control. The number of genes selected will influence the calculation time of the correlation.

Norm.interest <- NormData.log[rownames(select),]

dim(Norm.interest)
```


The "Norm.interest" object contains the normalized counts for DEGs. In order to build the gene co-expression network, the user needs to calculate correlation and correlation significance between each pair of DEGs in the whole experimental data set. There are several tools to calculate correlation, here we use the R package "psych" ( (https://www.rdocumentation.org/packages/psych/). Psych is a general-purpose toolbox originally developed for psychometrics analyses, but with useful tools for data analyses including correlation analysis.

The user can download the package from the Comprehensive R Archive Network (CRAN) using the following commands:

```{r message=FALSE}
#If it is not installed, install with:
#install.packages("mnormt")
install.packages("https://cran.r-project.org/src/contrib/Archive/psych/psych_1.7.5.tar.gz", repos=NULL, type="source")

library("psych")

```

Then, set the instructions for the correlation calculation for every DEGs pairs. Between the available methods select Pearson correlation to analyze the normalized data. The correlation results will be stored in "Norm.interest.corr" object:

```{r}
Norm.interest.corr <- corr.test( t(Norm.interest), method="pearson", ci=F)
```

Among the many results of this function, there are two triangular matrices with the needed data. One matrix is symmetric and contains the correlation values. The other matrix contains the p-values in the upper part and the adjusted p-values in the lower part. To generate a table comprising the data organized properly to visualize the network, the user should execute the following commands:


```{r}

Norm.interest.corr$p[lower.tri( Norm.interest.corr$p,diag=TRUE)]=NA
Pval.adj <- as.data.frame(as.table(Norm.interest.corr$p))
#The "Pval.adj" object contains all p-values from the lower part of the matrix.

Norm.interest.corr$r [lower.tri( Norm.interest.corr$r,diag=TRUE)]=NA
Correlation <- as.data.frame(as.table(Norm.interest.corr$r))

#The "Correlation" object contains all correlations from the matrix.

Cor.table <- na.exclude(cbind( Correlation, Pval.adj))[,c(1,2,3,6)]

#The "Cor.table" object contains all correlations and p-values selected from the matrix. The following command adds the names to the columns in "Cor.table":

colnames(Cor.table) <- c("gene1","gene2","cor","p.adj")

kable(head(Cor.table))
```

The generated "Cor.table" object, can be filtered based on absolute correlation (0.9) and adjusted p-value (0.01) thresholds:

```{r}
Cor.table.filt <- Cor.table [(abs(Cor.table[,3])>0.7 & Cor.table[,4] <0.01 ),]
write.table(Cor.table.filt, "Cor.table.filter.txt", sep="\t", row.names=F, quote=F)

dim(Cor.table.filt)
```



At this point, we generated the "Cor.table.filter.txt" file containing the statistically significant correlations across the whole data set for every pair of differentially expressed genes.


For this example, we will create a transcriptional network using a TF-Target database. We are going to cross the pairs that have a major correlation to 0.9 with the information obtained from the DAP-seq database (http://neomorph.salk.edu/dap_web/pages/browse_table_aj.php), which contains information from TF-TARGET pairs with evidence of possible binding.

Para este ejemplo utilizaremos un subset de esta base de datos, que la podemos descargar del github del curso (dapseq.subset.txt).


```{r}
dapseq<-read.table("dapseq.subset.txt.gz",header=T)
kable(head(dapseq))
```

Once loaded the TF-TARGET table  we intersect this table with the correlation data and write to a file.

```{r}

output.dapseq.filt<- unique(rbind(merge(dapseq,Cor.table.filt,by.x=c(1,2),by.y=c(1,2)),merge(dapseq,Cor.table.filt,by.x=c(1,2),by.y=c(2,1))))



write.table(output.dapseq.filt,"output.dapseq.filt.txt", sep="\t",quote=F,row.names = F)

kable(head(output.dapseq.filt))
```

With this instrucction we will identify the transcription factor used in the network.

```{r}
TFs<-unique(output.dapseq.filt[,1])
TFs<-cbind(TFs,"TF")
colnames(TFs)<-c("id","TF")
write.table(TFs,"TFs.txt",sep="\t",col.names=NA,quote=F)

```



#####  Network Visualization

Co-expression networks help associate genes that are involved in similar biological functions. The analysis and visualization of gene networks is a key and powerful step to identify relationships and discover important elements in the network. Analysis of gene networks also offers us the opportunity to formulate hypotheses about key genes and implicated biological functions. Below, we describe some simple steps to visualize a gene co-expression network.

Before we generate a network view, we will calculate a few useful network statistics. The user can later graphically represent network statistics on the same network. For instance, in Cytoscape (a popular software platform to view and analyze networks), statistics can be calculated and added as attributes to the nodes or edges. The number of connections of a node in a network is known as "degree", and it is a useful statistic to identify relevant nodes in biological networks, which are typically highly connected nodes or hubs. Another important node statistic is the number of times that a path passes through the node, which represents the amount of control that this node exerts over the interactions of other nodes in the networks known as "Betweenness centrality". To calculate network statistics in this guide we used "igraph" (http://igraph.org ). This R package can process simple graphs and network analysis handling large graphs if necessary. In addition, it offers functions for generating graph visualization and complete network statistics.

The basic statistics of the network, degree and betweenness, can be calculated using "igraph" R package. Download and install "igraph" with following commands:

```{r message=FALSE}
#If it is not installed, install with:
#install.packages("igraph")
library(igraph)
```


Select the columns 1 and 2 in the "Cor.table.filt" object. These columns contain all the DEGs pairs with high correlation. Store them in "g" object, which is the network:

```{r}
g <- graph.data.frame( output.dapseq.filt, directed=TRUE)
g
```


Calculate the degree and betweenness for each node in "g" object and store the results in the "degree" and "betweenness" objects respectively:

```{r}
degree <- degree(g)
betweenness <- betweenness(g)
Node_nw_st <- data.frame( degree, betweenness)

```


The "Node_nw_st" object contains all the calculated statistics for each node in "g" object. To integrate the two parameters, we generated two different ranking for degree and betweenness for all nodes in "Node_nw_st", and then calculated the mean of the rankings at each node. The nodes with the higher degree and betweenness will be scored with the higher value. The combined ranking can be added to the former table:

```{r}
Rank_stat <- rowMeans(cbind(rank(Node_nw_st[,1]), rank(Node_nw_st[,2])))
Node_nw_st <- cbind(Node_nw_st, Rank_stat)
write.table(Node_nw_st,file="Node_nw_st.txt", sep="\t", col.names = NA, quote=F)
kable(head(Node_nw_st))
```


The "Node_nw_st.txt" file contains all the calculated statistics for each node. These statistics will be used to complement the visualization of the "Cor.table.filter.txt" network created previously.

We will use Cytoscape to visualize the network and corresponding statistics. Cytoscape is an open source software platform. Cytoscape can be downloaded from http://www.cytoscape.org/ and requires JAVA™ JRE or JDK.

First, we launch Cytoscape, and then we import the network table "Cor.table.filter.txt". This can be done by selecting File > Import > Network > File. After selecting the file, the user should indicate where relevant information is stored in the file.
In our case, "Source Interaction" is in the first column which is labeled "gene1" and "Target Interaction" is in the second column which is labeled "gene2". These columns contain the gene ID information that Cytoscape will use to identify the nodes in each interacting pair in the network.

It is useful to keep the information contained in the table shown in the "Preview" window as an edge, (e.g. correlation value, adjusted p-values), the user should click the corresponding column header to activate it. After the network has been displayed, and to get a better visualization, nodes can be arranged following different layouts.

One of the most common displays is the "organic layout” which can be found under the Layout > yFiles Layout > Organic menu within Cytoscape. This layout will display nodes based in repulsive forces between the nodes and attractive forces induced by edges  (55), facilitating identification of highly connected nodes.

Since both degree and betweenness centrality are measures of the function of a node in network connectivity, we will use the combined ranking in the node attribute file "Node_nw_st.txt" generated early in this section  to map the size of the nodes to represent node importance, the file is located in the "WorkDir" folder.

To load the node information calculated before, the user should go through File > Import > Table > File and select "Node_nw_st.txt". In the pop-up window, the dropdown list "Network Collection", select the imported network, in this case "Cor.table.filter.txt". Import the table as "Node table columns" in the "Import Data as: " dropdown list and be sure all columns are checked. Node size can be set accordingly by selecting the "Style" tab under the "Control Panel" box. The "Style" tab also allows the user to set graphic properties on edges and the whole network in their corresponding sections in the lower part of the box.

Under the "Node" section in the "Style" tab of the "Control Panel" box, set the size of the node by checking "Lock node width and height", and then select "Size". On the new menu displayed, click on "Column" to show a new menu in which select "Rank_stat" as the attribute to determine the size. "Mapping type" should be in "Continuous Mapping" option to distribute the sizes continuously along the previously generated combined ranking for each node . The size variation can be adjusted in the graph that appears in the "Current Mapping" section.

The edges appearance can also be customized in a similar way to nodes. In the same "Style" tab in the "Control Panel", we select the "Edge" tab in the lower part of the section. Click the option "Edge color to arrows". In the option "Column" select "cor" as the mapping attribute. Then in "Mapping Type" select "Continuous Mapping" option. Clicking the graph is possible to adjust the colors and intensities. The user must set the minimal and maximum values every time, in this case -1 and 1 respectively.

#Performing cluster analysis of the network

Analysis of network topology or the structure that determine the way in which genes are connected is useful to derive biological insights. For instance, subnetworks of genes that are highly connected (network cluster) in a co-expression network are usually involved in similar biological functions. To find groups of genes that may be acting in a coordinated manner, the user can perform a cluster analysis of the network.

Numerous network clustering apps can be found at Cytoscape Application Store web (http:// apps.cytoscape.org/). For the purpose of this example, download, install and use "clusterMaker" plugin. clusterMaker offers many options to perform cluster analysis that can be addressed in its manual. We use "Community Clustering" (GLay) with the default options, because it provides an optimized layout for large networks and a structured visualization for more efficient exploration and analysis of biological networks . To display the clustering results as a new network we select "Create new clustered network", and we check "Assume edges are undirected". In this case, directionality cannot be assumed, because relationships were established based on correlation of gene expression. To keep the gene-gene connection we select "Restore inter-cluster edges after layout". This analysis generates a new network which contains the same nodes than the previous but arranged according to the connectivity of the nodes in the different clusters.

